src/platform/windows/audio.cpp

```diff
@@ -152,22 +152,22 @@ namespace platf::audio {
     wave_format.Format.wFormatTag = WAVE_FORMAT_EXTENSIBLE;
     wave_format.Format.nChannels = format.channels;
     wave_format.Format.nSamplesPerSec = SAMPLE_RATE;
-    wave_format.Format.wBitsPerSample = 16;
+    wave_format.Format.wBitsPerSample = 32;
     wave_format.Format.nBlockAlign = wave_format.Format.nChannels * wave_format.Format.wBitsPerSample / 8;
     wave_format.Format.nAvgBytesPerSec = wave_format.Format.nSamplesPerSec * wave_format.Format.nBlockAlign;
     wave_format.Format.cbSize = sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX);
 
-    wave_format.Samples.wValidBitsPerSample = 16;
+    wave_format.Samples.wValidBitsPerSample = 32;
     wave_format.dwChannelMask = format.channel_mask;
-    wave_format.SubFormat = KSDATAFORMAT_SUBTYPE_PCM;
+    wave_format.SubFormat = KSDATAFORMAT_SUBTYPE_IEEE_FLOAT;
 
     return wave_format;
   }
 
   int
   set_wave_format(audio::wave_format_t &wave_format, const format_t &format) {
     wave_format->nSamplesPerSec = SAMPLE_RATE;
-    wave_format->wBitsPerSample = 16;
+    wave_format->wBitsPerSample = 24;
```

Comment:
I'm concerned we might need some additional logic for this. I don't know if it's guaranteed that all devices will support 24-bit PCM for all channel configurations. I think it would be best if we tried them in order: 32-bit -> 24-bit -> 16-bit

That would ensure we don't possibly regress any audio devices that are 16-bit only, and also provide the highest quality on devices with 32-bit support.

Reply:
It's used only for "virtual" sinks, which I thought only consisted of steam virtual speakers. Am I wrong in this assumption?
```cpp
/**
 * If the requested sink is a virtual sink, meaning no speakers attached to
 * the host, then we can seamlessly set the format to stereo and surround sound.
 *
 * Any virtual sink detected will be prefixed by:
 *    virtual-(format name)
 * If it doesn't contain that prefix, then the format will not be changed
 */
std::optional<std::wstring>
set_format(const std::string &sink) {
  auto sink_info = get_sink_info(sink);

  // If the sink isn't a device name, we'll assume it's a device ID
  auto wstring_device_id = find_device_id_by_name(sink).value_or(from_utf8(sink_info.second.data()));

  if (sink_info.first == format_t::none) {
    // wstring_device_id does not contain virtual-(format name)
    // It's a simple deviceId, just pass it back
    return std::make_optional(std::move(wstring_device_id));
  }

  wave_format_t wave_format;
  auto status = policy->GetMixFormat(wstring_device_id.c_str(), &wave_format);
  if (FAILED(status)) {
    BOOST_LOG(error) << "Couldn't acquire Wave Format [0x"sv << util::hex(status).to_string_view() << ']';

    return std::nullopt;
  }

  set_wave_format(wave_format, formats[(int) sink_info.first - 1]);

  WAVEFORMATEXTENSIBLE p {};
  status = policy->SetDeviceFormat(wstring_device_id.c_str(), wave_format.get(), (WAVEFORMATEX *) &p);

  // Surround 5.1 might contain side-{left, right} instead of speaker in the back
  // Try again with different speaker mask.
  if (status == 0x88890008 && sink_info.first == format_t::surr51) {
    set_wave_format(wave_format, surround_51_side_speakers);
    status = policy->SetDeviceFormat(wstring_device_id.c_str(), wave_format.get(), (WAVEFORMATEX *) &p);
  }
  if (FAILED(status)) {
    BOOST_LOG(error) << "Couldn't set Wave Format [0x"sv << util::hex(status).to_string_view() << ']';

    return std::nullopt;
  }

  return std::make_optional(std::move(wstring_device_id));
}
```

Reply:
> It's used only for "virtual" sinks, which I thought only consisted of steam virtual speakers. Am I wrong in this assumption?

Other virtual speakers can be used, such as virtual audio cable https://vb-audio.com/Cable/

Reply:
Alright, so sunshine can have default or custom audio sink, and default or custom virtual audio sink.
Will check how it can be handled.

Reply:
I think it could just be as simple as a loop that tries `SetDeviceFormat()` until it succeeds.

Reply:
I thought it would be a simple loop too, but then ended up with a nasty memory corruption. The rest of the code had accumulated quite a bit of technical dept. Either way it should be ready soon.

Reply:
Pushed and tested what I could.